fix reno pre_release_tag_re pattern

brettlangdon opened the pull request:
## Commit Message
{{title}}

We have been having issues with how reno collapses pre-release notes into official release versions.

This change fixes a bug in how reno decides to collapse pre-release notes.

If we have the tags:

- v0.56.0
- v0.56.0rc7
- v0.56.0rc6
- ...

When reno iterates over these versions it will properly identify `v0.56.0rc7` as a pre-release, but then will identify `.0rc7` as the pre-release identify and strip it from the tag. So we end up with `v0.56` and it then will not collapse it because it cannot find a release for `v0.56`.

This change moves the `\v.\d+` pattern to outside the `pre_release` matching group in the regular expression so we can properly strip just `rc#` from the tag.

majorgreys commented:
Can you point to where this collapsing is visible?

On stable right now the following tags are broken out:

https://ddtrace.readthedocs.io/en/stable/release_notes.html

![image](https://user-images.githubusercontent.com/155673/145652059-44ce6854-329a-4984-9176-9ab3aa705699.png)

brettlangdon commented:
> Can you point to where this collapsing is visible?

this is where things get a little weird...

Reno will only ever look at a single branch. It will start with now, and then traverse backwards through the git history to find all the tags, and then also find all the release notes that have been added (moved/modified as well...?).

What this means is that because of our current branching strategy we have release tags whose commits only exist on the release branch and not on `master`.

So while `master` (generally) has all the release notes, `reno` cannot find the tag for `v0.56.0` since the commit `v0.56.0` is associated with only exists on the `0.56` branch.


This means we will not be able to see a benefit from these changes on `master` branch, but if you run `reno` on a specific release branch, then it'll work.



So, we can see the benefit if we run directly on the `0.56` branch.

<details>
<summary>Command run</summary>

```
reno report --branch=0.56 --no-show-source | \
  pandoc -f rst -t gfm --wrap=none | less
```

</details>

<details>
<summary>Release notes before this change</summary>

# Release Notes

## v0.56.1

### Bug Fixes

-   Fix error when calling `concurrent.futures.ThreadPoolExecutor.submit` with `fn` keyword argument.

## v0.56.0rc7

### Bug Fixes

-   The thread safety of the custom buffered encoder was fixed in order to eliminate a potential cause of decoding errors of trace payloads (missing trace data) in the agent.

## v0.56.0rc6

### Bug Fixes

-   Fix handling of Python exceptions during trace encoding. The tracer will no longer silently fail to encode invalid span data and instead log an exception.

## v0.56.0rc5

### Bug Fixes

-   Configure a writer thread in a child process after forking based on writer configuration from its parent process.

## v0.56.0rc3

### New Features

-   Profiling now supports tracing greenlets with gevent version prior to 1.3.
-   The heap profiler is now enabled by default.
-   Add yaaredis ≥ 2.0.0 support.

### Deprecation Notes

-   The `ddtrace.utils` module and all of its submodules have been copied over into `ddtrace.internal` in an effort to internalize these modules. Their public counterparts will be removed entirely in version 1.0.0.

### Bug Fixes

-   Allow the elasticsearch service name to be overridden using the integration config or the DD_SERVICE_MAPPING environment variable.
-   Fixes parsing of `botocore` env variables to ensure they are parsed as booleans.
-   Ensure tornado spans are marked as an error if the response status code is 500 \<= x \< 600.

## v0.56.0rc2

### Bug Fixes

-   Fix memory leak caused when the tracer is disabled.

## v0.56.0rc1

### Upgrade Notes

-   The aredis integration is now enabled by default.

### Deprecation Notes

-   The contents of `monkey.py` have been moved into `_monkey.py` in an effort to internalize the module. Public methods have been imported back into `monkey.py` in order to retain compatibility, but monkey.py will be removed entirely in version 1.0.0.

### Bug Fixes

-   Fix Pyramid caller_package level issue which resulted in crashes when starting Pyramid applications. Level now left at default (2).
-   Set the correct package name in the Pyramid instrumentation. This should fix an issue where the incorrect package name was being used which would crash the application when trying to do relative imports within Pyramid (e.g. when including routes from a relative path).

## v0.55.0rc1

### New Features

-   Add aredis support \>= 1.1.0
-   Add automatic unix domain socket detection for Dogstatsd. The expected path for the socket is `/var/run/datadog/dsd.socket` which if exists, will be used instead of the previous UDP default, `udp://localhost:8125/`. To be used in conjunction with `dogstatsd_socket` in your `datadog.yaml` file, or the `DD_DOGSTATSD_SOCKET` environment variable set on the Datadog agent.
-   Add new `DD_TRACE_SAMPLING_RULES` environment variable to override default sampling rules. For Example:: `DD_TRACE_SAMPLING_RULES='[{"sample_rate":0.5,"service":"my-service"}]'`
-   Add support for [snowflake-connector-python](https://pypi.org/project/snowflake-connector-python/) \>= 2.0.0. Note that this integration is in beta and is not enabled by default. See the snowflake integration documentation for how to enable.
-   Only for CI Visibility (`pytest` integration): include `pytest` version as a tag in the test span.
-   Added official support for Python 3.10
-   Only for CI Visibility (`pytest` integration): Extract stage and job name from environment data in Azure Pipelines.

### Upgrade Notes

-   Instead of using error constants from `ddtrace.ext.errors`. Use constants from `ddtrace.constants` module. For example: `ddtrace.ext.errors.ERROR_MSG` -> `ddtrace.constants.ERROR_MSG`
-   Instead of using priority constants from `ddtrace.ext.priority`. Use constants from `ddtrace.constants` module. For Example:: `ddtrace.ext.priority.AUTO_KEEP` -> `ddtrace.constants.AUTO_KEEP`
-   Instead of using system constants from `ddtrace.ext.system`. Use constants from `ddtrace.constants` module. For Example:: `ddtrace.ext.system.PID` -> `ddtrace.constants.PID`

### Deprecation Notes

-   Deprecate DATADOG_TRACE_AGENT_HOSTNAME, DATADOG_TRACE_AGENT_PORT, DATADOG_PRIORITY_SAMPLING, DATADOG_PATCH_MODULES in favor of their DD equivalents.

    \[Deprecated environment variable\] \| \[Recommended environment variable\]

    -   For `DATADOG_TRACE_AGENT_HOSTNAME`, use `DD_AGENT_HOST`
    -   For `DATADOG_TRACE_AGENT_PORT` use `DD_AGENT_PORT`
    -   For `DATADOG_PRIORITY_SAMPLING`, [follow ingestion controls](https://docs.datadoghq.com/tracing/trace_retention_and_ingestion/#recommended-setting-the-global-ingestion-rate-to-100)
    -   For `DATADOG_PATCH_MODULES`, use `DD_PATCH_MODULES`

-   Moved `ddtrace.ext.errors` constants into the `ddtrace.constants` module. `ddtrace.ext.errors` will be removed in v1.0. Shorthand error constant (MSG,TYPE,STACK) in `ddtrace.ext.errors` will be removed in v1.0. Function `get_traceback()` in ddtrace.ext.errors is now deprecated and will be removed v1.0.

-   Moved `ddtrace.ext.priority` constants into `ddtrace.constants` module.

-   Moved `ddtrace.ext.system` constants into `ddtrace.constants` module.

### Bug Fixes

-   Fixes an issue where all Django function middleware will share the same resource name.
-   Fixed an issue with gevent worker processes that caused them to crash and stop.
-   Fixes exceptions raised when logging during tracer initialization when `DD_LOGS_INJECTION` is enabled.
-   The `ddtrace.utils.wrappers.unwrap` function now raises an error if trying to unwrap a non-wrapped object.
-   Only for CI Visibility (`pytest` integration): Fix extraction of branch in GitLab CI.


**THE REST OF THE RELEASE NOTES**


</details>


<details>
<summary>Release notes after this change</summary>

# Release Notes

## v0.56.1

### Bug Fixes

-   Fix error when calling `concurrent.futures.ThreadPoolExecutor.submit` with `fn` keyword argument.

## v0.56.0

### New Features

-   Profiling now supports tracing greenlets with gevent version prior to 1.3.
-   The heap profiler is now enabled by default.
-   Add yaaredis ≥ 2.0.0 support.

### Upgrade Notes

-   The aredis integration is now enabled by default.

### Deprecation Notes

-   The `ddtrace.utils` module and all of its submodules have been copied over into `ddtrace.internal` in an effort to internalize these modules. Their public counterparts will be removed entirely in version 1.0.0.
-   The contents of `monkey.py` have been moved into `_monkey.py` in an effort to internalize the module. Public methods have been imported back into `monkey.py` in order to retain compatibility, but monkey.py will be removed entirely in version 1.0.0.

### Bug Fixes

-   Allow the elasticsearch service name to be overridden using the integration config or the DD_SERVICE_MAPPING environment variable.
-   Fixes parsing of `botocore` env variables to ensure they are parsed as booleans.
-   Fix memory leak caused when the tracer is disabled.
-   The thread safety of the custom buffered encoder was fixed in order to eliminate a potential cause of decoding errors of trace payloads (missing trace data) in the agent.
-   Fix handling of Python exceptions during trace encoding. The tracer will no longer silently fail to encode invalid span data and instead log an exception.
-   Ensure tornado spans are marked as an error if the response status code is 500 \<= x \< 600.
-   Configure a writer thread in a child process after forking based on writer configuration from its parent process.
-   Fix Pyramid caller_package level issue which resulted in crashes when starting Pyramid applications. Level now left at default (2).
-   Set the correct package name in the Pyramid instrumentation. This should fix an issue where the incorrect package name was being used which would crash the application when trying to do relative imports within Pyramid (e.g. when including routes from a relative path).


</details>


It is also worth noting the difference that with this change, `reno` properly stops at the correct base release version. e.g. when running on `0.56` branch it only generates release notes for `v0.56.x`, but as it is right now it'll give you all the notes.


This change alone won't fix the issues with the way release notes are getting rendered in our docs, but it does fix a bug to properly identify and collapse pre-release tags.

jd commented:
One way of fixing that is to do a dummy git merge. Once v0.56.x is tagged, you merge this tag into `master` so it has it.
Your tree would be a bit weird, but since we're only doing backports, there should be no conflict and then the history would work as you expect:
![CleanShot 2021-12-24 at 10 08 39](https://user-images.githubusercontent.com/2644/147338387-f5bb37a3-0231-4f18-bb1e-310a52d3a712.png)
